File: Dsa-day86.py
from collections import Counter
from typing import List

class Solution:
    def maximumTotalDamage(self, power: List[int]) -> int:
        if not power:
            return 0

        # Calculate total damage for each unique power value
        counts = Counter(power)
        
        # Get sorted unique power values
        unique_power = sorted(counts.keys())
        n = len(unique_power)

        # Pre-calculate the total damage if all spells of a certain power are cast
        # total_damage_map[p] = p * count[p]
        total_damage_map = {p: p * counts[p] for p in unique_power}

        # dp[i] is the maximum total damage considering spells with damage up to unique_power[i]
        # Use a list for DP array
        dp = [0] * n

        # Base case: i = 0
        dp[0] = total_damage_map[unique_power[0]]

        # DP transition
        for i in range(1, n):
            current_p = unique_power[i]
            current_damage = total_damage_map[current_p]

            # Choice 1: Skip all spells with damage current_p
            skip_damage = dp[i - 1]

            # Choice 2: Cast all spells with damage current_p
            cast_damage = current_damage

            # Find the largest index j < i such that unique_power[j] < current_p - 2
            # This is the max damage from a non-conflicting subset
            j = i - 1
            while j >= 0 and unique_power[j] >= current_p - 2:
                j -= 1
            
            if j >= 0:
                cast_damage += dp[j]
            
            dp[i] = max(skip_damage, cast_damage)

        return dp[n - 1]

File: Dsa-day86.java
import java.util.*;

class Solution {
    public long maximumTotalDamage(int[] power) {
        if (power == null || power.length == 0) {
            return 0;
        }

        // 1. Count frequencies and calculate total damage for each unique power
        Map<Integer, Long> totalDamageMap = new HashMap<>();
        for (int p : power) {
            totalDamageMap.put(p, totalDamageMap.getOrDefault(p, 0L) + (long)p);
        }

        // 2. Get sorted unique power values
        List<Integer> uniquePower = new ArrayList<>(totalDamageMap.keySet());
        Collections.sort(uniquePower);

        int n = uniquePower.size();
        if (n == 0) {
            return 0;
        }

        // 3. Dynamic Programming
        // dp[i] is the maximum total damage considering spells with damage up to uniquePower.get(i)
        long[] dp = new long[n];

        // Base case: i = 0
        dp[0] = totalDamageMap.get(uniquePower.get(0));

        for (int i = 1; i < n; ++i) {
            int currentP = uniquePower.get(i);
            long currentDamage = totalDamageMap.get(currentP);

            // Choice 1: Skip all spells with damage currentP
            long skipDamage = dp[i - 1];

            // Choice 2: Cast all spells with damage currentP
            long castDamage = currentDamage;

            // Find the largest index j < i such that uniquePower.get(j) < currentP - 2
            int j = i - 1;
            while (j >= 0 && uniquePower.get(j) >= currentP - 2) {
                j--;
            }

            if (j >= 0) {
                castDamage += dp[j];
            }

            dp[i] = Math.max(skipDamage, castDamage);
        }

        return dp[n - 1];
    }
}